src/task_impl/atomic_task.rs

```diff
@@ -86,90 +73,96 @@ impl AtomicTask {
-                debug_assert!(state != LOCKED_WRITE_NOTIFIED, "unexpected state LOCKED_WRITE_NOTIFIED");
-
-                // Currently in a read locked state, this implies that `notify`
-                // is currently being called on the old task handle. So, we call
-                // notify on the new task handle
+            NOTIFYING => {
+                // Currently in the process of notifying the task, i.e.,
+                // `notify` is currently being called on the old task handle.
+                // So, we call notify on the new task handle
                 task.notify();
             }
+            _ => {
+                // TODO: Assert valid state
+            }
         }
     }
 
     /// Notifies the task that last called `register`.
     ///
     /// If `register` has not been called yet, then this does nothing.
     pub fn notify(&self) {
-        let mut curr = WAITING;
-
-        loop {
-            if curr == LOCKED_WRITE {
-                // Transition the state to LOCKED_NOTIFIED
-                let actual = self.state.compare_and_swap(LOCKED_WRITE, LOCKED_WRITE_NOTIFIED, Release);
-
-                if curr == actual {
-                    // Success, return
-                    return;
-                }
-
-                // update current state variable and try again
-                curr = actual;
-
-            } else if curr == LOCKED_WRITE_NOTIFIED {
-                // Currently in `LOCKED_WRITE_NOTIFIED` state, nothing else to do.
-                return;
-
-            } else {
-                // Currently in a LOCKED_READ state, so attempt to increment the
-                // lock count.
-                let actual = self.state.compare_and_swap(curr, curr + 1, Acquire);
-
-                // Locked acquired
-                if actual == curr {
-                    // Notify the task
-                    unsafe {
-                        if let Some(ref task) = *self.task.get() {
-                            task.notify();
-                        }
-                    }
+        // AcqRel ordering is used in order to acquire the value of the `task`
+        // cell as well as to establish a `release` ordering with whatever
+        // memory the `AtomicTask` is associated with.
+        match self.state.fetch_or(NOTIFYING, AcqRel) {
+            WAITING => {
+                // The notifying lock has been acquired.
+                let task = unsafe { (*self.task.get()).take() };
 
-                    // Release the lock
-                    self.state.fetch_sub(1, Release);
+                // Release the lock
+                self.state.fetch_and(!NOTIFYING, Release);
 
-                    // Done
-                    return;
+                if let Some(task) = task {
+                    task.notify();
                 }
-
-                // update current state variable and try again
-                curr = actual;
-
+            }
+            _ => {
```

Comment:
Nit: can you list out the remaining cases, just to make it clear that all of them should indeed be no-ops?

(REGISTERING, NOTIFYING, REGISTERING | NOTIFYING)